### ISSUE TITLE
Inconsistent Item class causing an obscure bug

### ISSUE THREAD

**Issue by jimtng** [boc-tothefuture/openhab-jruby#48]
It seems that some item type is monkeypatched, and some aren't. For example, SwitchItem is monkeypatched whereas StringItem is delegated / not monkeypatched.

This inconsistency caused a problem/bug. 

Given the following items file:
```
Switch SwitchItem1
Switch SwitchItem2
String StringItem1
```

At first I implemented the following rule

```
ITEMS_MAP = { SwitchItem1 => "X", SwitchItem2 => "Y" }
ITEMS = ITEMS_MAP.keys

rule 'x' do
  changed ITEMS
  triggered do |item|
    mapped_value = ITEMS_MAP[item] 
  end
end
```

This works well, because `item` passed by `triggered` is of the same class: `Java::OrgOpenhabCoreLibraryItems::SwitchItem` as the keys in my LIGHTS_MAP, and this is the expected behaviour.

However, when I have a StringItem as my key, it falls apart because an item `StringItem1` is not Java::OrgOpenhabCoreLibraryItems::StringItem, but instead it is `OpenHAB::Core::DSL::Items::StringItem`, so it will never match the triggered item.

Code sample (will not work as expected):
```
ITEMS_MAP = { StringItem1 => "X" }
ITEMS = ITEMS_MAP.keys

rule 'x' do
  changed ITEMS
  triggered do |item|
    mapped_value = ITEMS_MAP[item] # always returns nil
  end
end
```

For this reason, perhaps we should monkeypatch all item types?

**Comment by boc-tothefuture** [comment]
I was moving away from monkeypatching towards delegation because of issues with monkeypatching I was seeing.  So are you saying in general right now StringItems can't be used as triggers? 

I think (and it was in another thread around the restore states) every one of these may need a "to_oh" method that returns the delegated object, we we need for groups, etc before we pass it back into the framework. 

One issue that pushed me away from monkeypatching into delegation was the NumberItem, I can't monkeypatch that one because if I don't inherit from Numeric that ruby can't do math on it.  So we will always have a mixed bag.

**Comment by jimtng** [comment]
> So are you saying in general right now StringItems can't be used as triggers?

StringItems can be used as triggers. The problem was that triggered { |item| } gives me the Java OpenHAB StringItem, whereas within JRuby, StringItem is not the Java OpenHAB StringItem. So the ITEMS_MAP won't match the key.

**Comment by jimtng** [comment]
The other alternative is that all item types have their own jruby class, much like GroupItem and StringItem. However, in order to make my special case to work, the "item" as passed by run / triggered should also match to the same type, and not just the plain java type. Either way it will be consistent.

**Comment by boc-tothefuture** [comment]
Certainly for some types (numbers, etc) they can't be the Java version.  We can wrap those types coming in from rules to be the ruby versions.  That isn't a problem.

**Comment by jimtng** [comment]
> Certainly for some types (numbers, etc) they can't be the Java version. We can wrap those types coming in from rules to be the ruby versions. That isn't a problem.

If that were the case, then we will need to wrap all types, including the items that are sent by `run` and `triggered`. It is just for consistency so that the item type in the conditional blocks (changed, updated, received_command) match the item type in the run/triggered.

**Comment by jimtng** [comment]
Closing this because it will be resolved when #58 is resolved.
